web-api/src/storage/elastic.rs

```diff
@@ -339,3 +331,181 @@ struct IngestedDocument<'a> {
+            "must_not": [
+                excluded_ids,
+                {
+                    "range": {
+                        "properties.publication_date": {
+                            "gt": time
+                        }
+                    }
+                }
+            ]
+        })
+    }
+}
+
+fn create_knn_request_object(
+    embedding: &NormalizedEmbedding,
+    count: usize,
+    num_candidates: usize,
+    filter: &JsonObject,
+) -> JsonObject {
+    // https://www.elastic.co/guide/en/elasticsearch/reference/current/search-search.html
+    json_object!({
+        "knn": {
+            "field": "embedding",
+            "query_vector": embedding,
+            "k": count,
+            "num_candidates": num_candidates,
+            "filter": {
+                "bool": filter
+            }
+        }
+    })
+}
+
+fn normalize_scores<K>(mut scores: HashMap<K, f32>) -> HashMap<K, f32>
+where
+    K: Eq + Hash,
+{
+    let max_score = scores
+        .values()
+        .max_by(|l, r| l.total_cmp(r))
+        .copied()
+        .unwrap_or_default();
+    // TODO[pmk/now] the original code did a .max(1.0); this looked like bug, verify
+
+    for score in scores.values_mut() {
+        *score /= max_score;
+    }
+
+    scores
+}
+
+fn merge_scores_weighted<K>(
+    scores: impl IntoIterator<Item = (f32, HashMap<K, f32>)>,
+) -> HashMap<K, f32>
+where
+    K: Eq + Hash,
+{
+    // TODO[pmk/now] this originally didn't apply weights to scores where there was only one hit
+    //               this seemed like a bug and was removed, verify if that is correct
+    let weighted = scores.into_iter().flat_map(|(weight, mut scores)| {
+        for score in scores.values_mut() {
+            *score *= weight;
+        }
+        scores
+    });
+    collect_summing_repeated(weighted)
+}
+
+/// Reciprocal Rank Fusion
+#[allow(dead_code)]
+fn rrf<K>(k: f32, scores: impl IntoIterator<Item = HashMap<K, f32>>) -> HashMap<K, f32>
+where
+    K: Eq + Hash,
+{
+    let rrf_scores = scores.into_iter().flat_map(|scores| {
+        scores
+            .into_iter()
+            .sorted_by(|l, r| l.1.total_cmp(&r.1).reverse())
+            .enumerate()
+            .map(|(rank0, (document, _))| {
+                #[allow(clippy::cast_precision_loss)]
+                (document, (k + rank0 as f32).recip())
+            })
+    });
+    collect_summing_repeated(rrf_scores)
+}
+
+fn collect_summing_repeated<K>(scores: impl IntoIterator<Item = (K, f32)>) -> HashMap<K, f32>
+where
+    K: Eq + Hash,
+{
+    scores
+        .into_iter()
+        .fold(HashMap::new(), |mut acc, (key, value)| {
+            acc.entry(key).or_default().add_assign(value);
+            acc
+        })
+}
+
+fn take_highest_n_scores<K>(n: usize, scores: HashMap<K, f32>) -> HashMap<K, f32>
+where
+    K: Eq + Hash,
+{
+    scores
+        .into_iter()
+        .sorted_unstable_by(|(_, s1), (_, s2)| s1.total_cmp(s2).reverse())
+        .take(n)
+        .collect()
```

Comment:
:art: we should return early if `scores.len() <= n`, otherwise we sort and allocate in vain.